src/views/PianoRoll/Main.lua

```diff
@@ -0,0 +1,118 @@
+        input[k] = v
+    end
+    input.joy = {}
+    CloneInto(input.joy, joypad.get(1))
+
+    if TASState.movement_mode == MovementModes.disabled then
+        input.movement_mode = MovementModes.manual
+        input.manual_joystick_x = Joypad.input.X
+        input.manual_joystick_y = Joypad.input.Y
+    end
+    input.preview_joystick_x = input.manual_joystick_x
+    input.preview_joystick_y = input.manual_joystick_y
+end
+
+local SelectionGui = dofile(views_path .. "PianoRoll/SelectionGui.lua")
+local FrameListGui = dofile(views_path .. "PianoRoll/FrameListGui.lua")
+local JoystickGui = dofile(views_path .. "PianoRoll/JoystickGui.lua")
+local ToolsGui = dofile(views_path .. "PianoRoll/ToolsGui.lua")
+
+emu.atupdatescreen(function()
+    -- prevent reentrant calls caused by GUI actions while the game is running
+    if PianoRollContext.current == nil or PianoRollContext.current then return end
+
+    PianoRollContext.current:update()
+end)
+
+---public API---
+
+---Retrieves a TASState as determined by the currently active piano roll for the current frame identified by the current global timer value.
+---
+---If the current piano roll does not define what to do for this frame, or there is no current piano roll, nil is returned instead.
+---
+---@return table|nil override A table that can be assigned to TASState, additionally holding a field 'joy' that can be passed to joypad.set(...).
+function CurrentPianoRollOverride()
+    if (PianoRollContext.current == nil) then return nil end
+    local globalTimer = GetGlobalTimer()
+    if PianoRollContext.current ~= nil and globalTimer >= PianoRollContext.current.endGT then
+        local input = {}
+        RecordPianoRollInput(input)
+        PianoRollContext.current.endGT = globalTimer + 1
+        PianoRollContext.current.previewGT = globalTimer
+        PianoRollContext.current.editingGT = globalTimer
+        PianoRollContext.current.frames[globalTimer] = input
+    end
+
+    return PianoRollContext.current.frames[globalTimer]
+end
+
+---@class PianoRollContext
+---@field public current PianoRoll|nil The currently selected and active piano roll.
+---@field public all table An array of up to 7 piano rolls to easily switch between for rapid iteration, indexed from 1 through 7 inclusively.
+---@field public maxDisplayedFrames integer The maximum number of frames to display at once.
+PianoRollContext = {
+    current = nil,
+    all = {},
+    maxDisplayedFrames = 15,
+
+    ---Retrieves the current piano roll, raising error when it is nil
+    ---@return PianoRoll current The current PianoRoll, never nil
+    AssertedCurrent = function()
+        if PianoRollContext.current == nil then
+            error("Expected PianoRollContext.current to not be nil.", 2)
+        end
+        return PianoRollContext.current
+    end,
+}
+
+return {
+    name = "Piano Roll",
+    draw = function()
+        -- hack to get a denser UI
+        local previousGridGap = Settings.grid_gap
```

Comment:
Well duh, it's explicitly _explained_ to be a hack :)  
I'm not seeing this "other comment" you're referring to here, and what bug specifically does this cause?

Reply:
The other comment is on Discord

> the main issue is that the design is way too cramped and, as a consequence, encourages magic numbers and layout hacks/weirdness

and not particularly helpful 😅. Point being, the UI shouldn't be turned into a bunch of compacted blocks via a hack just because controls would be too small otherwise. That's an issue which must be solved one level higher, at the actual design stage.

As for the bug, restarting while on the piano roll page persists the 0 grid gap.

Reply:
Ah okay, I didn't notice the bug - particularly _because_ it persisted, and the grid gap being 0 didn't bother me at all, so I didn't even notice that happened.

I don't know how the UI should be designed to be "less cramped" - really I think of it as "less wasteful" - but I can't help but feel like this kind of tool _needs_ as much space as it can get. That's why I went for the hack in the first place.  

Could there perhaps be a way to define some `grid_rect` function that accepts an explicit `gap` parameter, so that certain controls that "need" to be cramped (e.g. the entirety of the joystick section) could just pass that cleanly?
 
I'd honestly prefer if this kind of tool could live in its own window, but that's a different beast entirely.

Reply:
I have implemented this in 254306c3a09b0d4d5b549a5216305c35204039a8, and used it where appropriate. Regular buttons and the like have been reorganized to use the space more efficiently.

Reply:
Looks good 👍